client/src/nv_ingest_client/message_clients/rest/rest_client.py

```diff
@@ -0,0 +1,219 @@
+                url = f"http://{self._host}:{self._port}{self._fetch_endpoint}/{job_state.job_spec.job_id}"
+                result = requests.get(url)
+                logger.debug(f"Fetch Message submitted to http endpoint {self._submit_endpoint}")
+                
+                # If the result contains a 200 then return the raw JSON string response
+                if result.status_code == 200:
+                    return result.text
+                else:
+                    # Follow the established backoff approach
+                    retries += 1
+                    backoff_delay = min(2**retries, self._max_backoff)
+                    
+                    if self.max_retries > 0 and retries <= self.max_retries:
+                        logger.error(f"Fetch attempt failed, retrying in {backoff_delay}s...")
+                        time.sleep(backoff_delay)
+                    else:
+                        logger.error(f"Failed to fetch message from {job_state.response_channel} after {retries} attempts.")
+                        raise ValueError(f"Failed to fetch message from HTTP endpoint after {retries} attempts: {err}")
+
+            except httpx.HTTPError as err:
+                retries += 1
+                logger.error(f"REST error during fetch: {err}")
+                backoff_delay = min(2**retries, self._max_backoff)
+
+                if self.max_retries > 0 and retries <= self.max_retries:
+                    logger.error(f"Fetch attempt failed, retrying in {backoff_delay}s...")
+                    time.sleep(backoff_delay)
+                else:
+                    logger.error(f"Failed to fetch message from {job_state.response_channel} after {retries} attempts.")
+                    raise ValueError(f"Failed to fetch message from HTTP endpoint after {retries} attempts: {err}")
+
+                # Invalidate client to force reconnection on the next try
+                self._client = None
+            except Exception as e:
+                # Handle non-http specific exceptions
+                logger.error(f"Unexpected error during fetch from {job_state.response_channel}: {e}")
+                raise ValueError(f"Unexpected error during fetch: {e}")
+
+    def submit_message(self, _: str, message: JobSpec) -> str:
+        """
+        Submits a message to a specified HTTP endpoint with retries on failure.
+
+        Parameters
+        ----------
+        channel_name : str
+            Not used as part of RestClient
+        message : str
+            The message to submit.
+
+        Raises
+        ------
+        RedisError
+            If submitting the message fails after the specified number of retries.
+        """
+        retries = 0
+        while True:
+            try:
+                # Submit via HTTP
+                url = f"http://{self._host}:{self._port}{self._submit_endpoint}"
+                result = requests.post(url, data=message.json(), headers={"Content-Type": "application/json"})
```

Comment:
I need to discuss with Devin what is actually a "terminal failure" vs "retry failure" once we discuss that obviously here depending on the results.status_code either an exception needs to be thrown with a clear error or the standard retry logic should be attempted.

Right now regardless of what status_code is returned from the server the client just assumed everything is fine which is obviously not good.

Reply:
Added a common approach of hard failing on most 4xx responses and some 5xx responses. All other response codes will result in a retry attempt.